### PR TITLE
Clean up factory/interpreter: split importing code

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Import work prep

- Remove duplication of imported modules between factory and interpreter
- Move resolution code into a separate class
- Move caching code into its own class

Otherwise no logic change

### CODE REVIEW

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonInterpreter.cs [hunk 1]
```diff
@@ -24,39 +24,46 @@
...
-        private readonly ConcurrentDictionary<string, IPythonModule> _modules;
-        private readonly AstPythonBuiltinType _noneType;
 
-        internal readonly AnalysisLogWriter _log;
-
-        private readonly object _userSearchPathsLock = new object();
         private IReadOnlyList<string> _userSearchPaths;
         private IReadOnlyDictionary<string, string> _userSearchPathPackages;
-        private HashSet<string> _userSearchPathImported;
+        private ConcurrentBag<string> _userSearchPathImported = new ConcurrentBag<string>();
```
Comment (MikhailArkhipov) [inline src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonInterpreter.cs]: В принципе поскольку lock  я вернул, можно вернуть Hashset

File: src/Analysis/Engine/Impl/Interpreter/Definitions/IPythonInterpreterFactory.cs [hunk 2]
```diff
@@ -47,4 +47,11 @@ public interface IPythonInterpreterFactory {
         /// </summary>
         void NotifyImportNamesChanged();
     }
+
+    public interface IPythonInterpreterFactory2: IPythonInterpreterFactory {
```
Comment (AlexanderSher) [inline src/Analysis/Engine/Impl/Interpreter/Definitions/IPythonInterpreterFactory.cs]: Is it implemented anywhere?

Comment (MikhailArkhipov) [reply src/Analysis/Engine/Impl/Interpreter/Definitions/IPythonInterpreterFactory.cs]: `AstPythonInterpreterFactory`

File: src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonInterpreter.cs (hunk 1 above)
Comment (AlexanderSher) [reply src/Analysis/Engine/Impl/Interpreter/Ast/AstPythonInterpreter.cs]: Yep

**Pull request merged**